language/stdlib/modules/libra_system.move

```diff
@@ -186,69 +196,186 @@ module LibraSystem {
+            &validator_set_ref.validators,
+            account_address
+        );
+
+        // remove corresponding ValidatorInfo from the validator set
+        _  = Vector::swap_remove(
+             &mut validator_set_ref.validators,
+            to_remove_index
+        );
+        _  = Vector::swap_remove(
+            &mut discovery_set_ref.discovery_set,
+            to_remove_index
+        );
+
+        reconfigure_();
+        emit_discovery_set_change();
+    }
+
+    public fun rotate_consensus_pubkey(consensus_pubkey: vector<u8>) acquires ValidatorSet {
+        let validator_set_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+        let account_address = Transaction::sender();
+
+        // Ensure that this address is already a validator
         Transaction::assert(
-            !Vector::contains(
-                &validator_set_ref.removals,
-                &account_address,
-            ),
-            22
+            is_validator_(&account_address, &validator_set_ref.validators),
+            21
         );
 
-        // Add to removals
-        Vector::push_back(
-            &mut validator_set_ref.removals,
-            account_address,
+        ValidatorConfig::rotate_consensus_pubkey(consensus_pubkey);
+
+        let validator_index = get_validator_index(
+            &validator_set_ref.validators,
+            account_address
+        );
+
+        if(copy_validator_info(Vector::borrow_mut(&mut validator_set_ref.validators, validator_index))) {
+            reconfigure_();
+        }
+    }
+
+    public fun rotate_validator_network_identity_pubkey(
+        validator_network_identity_pubkey: vector<u8>
+    ) acquires DiscoverySet {
+        let discovery_set_ref = borrow_global_mut<DiscoverySet>(0xD15C0);
+        let account_address = Transaction::sender();
+
+        ValidatorConfig::rotate_validator_network_identity_pubkey(validator_network_identity_pubkey);
+
+        let validator_index = get_discovery_index(
+            &discovery_set_ref.discovery_set,
+            account_address
+        );
+
+        if(copy_discovery_info(Vector::borrow_mut(&mut discovery_set_ref.discovery_set, validator_index))) {
+            emit_discovery_set_change();
+        }
+    }
+
+    public fun rotate_validator_network_address(
+        validator_network_address: vector<u8>
+    ) acquires DiscoverySet {
+        let discovery_set_ref = borrow_global_mut<DiscoverySet>(0xD15C0);
+        let account_address = Transaction::sender();
+
+        ValidatorConfig::rotate_validator_network_address(validator_network_address);
+
+        let validator_index = get_discovery_index(
+            &discovery_set_ref.discovery_set,
+            account_address
+        );
+
+        if(copy_discovery_info(Vector::borrow_mut(&mut discovery_set_ref.discovery_set, validator_index))) {
+            emit_discovery_set_change();
+        }
+    }
+
+    fun reconfigure_() acquires ValidatorSet {
```

Comment:
I think we discussed this, but to help me remember: what goes wrong if you call this `reconfigure_` from the genesis?

Reply:
The problem is that the block time is set to zero initially. If we call `reconfigure_` here we will have an assertion error due to the block time constraint. Setting the block time to 1 initially doesn't work for some reasons as well. I guess it has something to do with the default value for timeout block used by the consensus.

Reply:
> Setting the block time to 1 initially doesn't work for some reasons as well. I guess it has something to do with the default value for timeout block used by the consensus.

This seems important to dig into. We should understand what expectations consensus has about the time value, and it would be nice not to have to split the code like this.

Reply:
I just talked to @zekun000 and can confirm that the genesis block will have the timestamp of 0. That's the reason why we can't set the time ourselves.

Reply:
the timestamp in LedgerInfo::genesis is just a const(currently 0), but it's fine as long as it's consistent with the on-chain state.